File: src/fell_finder/containers/graph_data.py
"""Containers for the data to be stored against graph nodes/edges"""

from typing import Any, Dict


class GraphNode:
    """Container for additional node attributes which are not supported
    natively by rustworkx"""

    def __init__(self, node_dict: Dict[str, Any]) -> None:
        self.index = None

        self.lat = node_dict["lat"]
        self.lon = node_dict["lon"]
        self.elevation = node_dict["elevation"]

        self.dist_to_start = None

    def set_index(self, index: int) -> None:
        """Set the node index to the provided value"""
        self.index = index

    def set_dist_to_start(self, dist: float) -> None:
        """Set the distance from the node to the start point"""
        self.dist_to_start = dist


class GraphEdge:
    """Container for additional edge attributes which are not supported
    natively by rustworkx"""

    def __init__(self, edge_dict: Dict[str, Any]) -> None:
        self.index = None

        self.highway = edge_dict["highway"]
        self.surface = edge_dict["surface"]
        self.distance = edge_dict["distance"]
        self.elevation_gain = edge_dict["elevation_gain"]
        self.elevation_loss = edge_dict["elevation_loss"]
        self.geometry = edge_dict["geometry"]

    def set_index(self, index: int) -> None:
        """Set the edge index to the provided value"""
        self.index = index

File: packages/fell_finder/src/common/graph_data.rs
/// Sets the data which will be stored as weights in the petgraph graph, all
/// attributes will be populated after the graph is created
#[derive(Default, Debug, Clone, Copy, PartialEq)]
pub struct NodeData {
    pub id: i64,
    pub lat: f64,
    pub lon: f64,
    pub elevation: f64,
    pub is_start: bool,
    pub dist_to_start: Option<f64>,
}

/// Container for edge metadata which will be stored in the graph, all
/// attributes are populated based on the output of a SQL query
#[derive(Default, Debug, Clone, PartialEq)]
pub struct EdgeData {
    pub src: i64,
    pub dst: i64,
    pub highway: String,
    pub surface: String,
    pub elevation_gain: f64,
    pub elevation_loss: f64,
    pub distance: f64,
    pub lats: Vec<f64>,
    pub lons: Vec<f64>,
    pub eles: Vec<f64>,
    pub dists: Vec<f64>,
}
